### PR TITLE
Threading API conversion, vol. 2

### PR DESCRIPTION
**Pull request by ghedo**
A few more conversions, we are almost there!!

But the "big ones" (`md_rand,c`, `mem_dbg.c`, `engines/`, RSA blinding) still need conversion and will likely take some more time (not sure if I can do this in time for the freeze, help would be appreciated :) ).

### CODE REVIEW

**Comment by ghedo** [comment]
Interestingly there are a bunch of locks that are currently unused:
```
# define CRYPTO_LOCK_ECDSA               32
# define CRYPTO_LOCK_ECDH                34
# define CRYPTO_LOCK_BN                  35
# define CRYPTO_LOCK_STORE               37
# define CRYPTO_LOCK_COMP                38
# define CRYPTO_LOCK_FIPS                39
# define CRYPTO_LOCK_FIPS2               40
```

File: crypto/bio/b_addr.c [hunk 1]
```diff
@@ -55,12 +55,16 @@
...
 #include "bio_lcl.h"
+#include "internal/threads.h"
 
 #ifndef OPENSSL_NO_SOCK
 #include <openssl/err.h>
 #include <openssl/buffer.h>
 #include <ctype.h>
 
+CRYPTO_RWLOCK *bio_lookup_lock;
+CRYPTO_ONCE bio_lookup_init = CRYPTO_ONCE_STATIC_INIT;
```
Comment (ajmohan) [inline crypto/bio/b_addr.c]: both may be defined as static.

Comment (ghedo) [inline crypto/bio/b_addr.c]: Fixed, thanks.

**Comment by mattcaswell** [comment]
+1

**Comment by mattcaswell** [comment]
@ghedo, have you started: md_rand,c, mem_dbg.c, engines/, RSA blinding? I can maybe take some of this on.

**Comment by mattcaswell** [comment]
@ghedo, I took a gamble and I have a patch for md_rand.c

**Comment by ghedo** [comment]
@mattcaswell I only looked at RSA blinding but I haven't finished yet. I sort of wanted to convert that code to something like what BoringSSL does (which is saner IMO) but I guess it's too late for that at this point:
https://boringssl.googlesource.com/boringssl/+/master/crypto/rsa/rsa_impl.c#212

**Comment by ghedo** [comment]
I just fixed a couple of memory leaks: the RSA lock wasn't freed properly and I implemented the `CRYPTO_MEM_CHECK_DISABLE` dance for `ex_data_lock` (I don't think we ever want to free that, but maybe the init code could be used for this? I don't really know how that works).

**Comment by mattcaswell** [comment]
Still +1...ping @richsalz.

**Comment by richsalz** [comment]
will merge.  @mattcaswell you should post your branch here for @ghedo to look at.

**Comment by mattcaswell** [comment]
You mean the one you just +1'd?? How about I just merge it :-)

**Comment by richsalz** [comment]
Please do ☺
Feel free to merge anything that’s ready, I was just picking up the slack for others with more important stuff.

**Comment by richsalz** [comment]
sorry, not enough email context.  I meant that @ghedo might have useful feedback on what I just plus-one'd.

And I merged this and am closing it. :)

**Pull request closed without merging**